RTClib.cpp

```diff
@@ -746,6 +746,130 @@ TimeSpan::TimeSpan(int16_t days, int8_t hours, int8_t minutes, int8_t seconds)
 /**************************************************************************/
 TimeSpan::TimeSpan(const TimeSpan &copy) : _seconds(copy._seconds) {}
 
+/*!
+    @brief  Create a new TimeSpan from an iso8601 formatted period string
+
+    If the string is entirely malformed (doesn't begin with P or -) this will
+    construct a TimeSpan of 0 seconds. If parsing fails before the end of the
+    string, this constuctor will interpret its argument as the longest
+    well-formed prefix string.
+
+    For example, the string P5M10~ will parse the same as P5M
+
+    @param iso8601 the formatted string
+*/
+TimeSpan::TimeSpan(char* iso8601): _seconds(0) {
+  int32_t negative = 1;
+  char* cursor = iso8601;
+  if (*cursor != '\0' && *cursor == '-') {
+    negative = -1;
+    cursor++;
+  }
+  if (*cursor != '\0' && *cursor == 'P') {
+    cursor++;
+  } else {
+    return;
+  }
+
+  char* rest;
+  long num = strtol(cursor, &rest, 10);
+  cursor = rest;
+  if (*cursor != '\0' && *cursor == 'D') {
+    _seconds += negative * SECONDS_PER_DAY*num;
+    cursor++;
+  }
+
+  if (*cursor != '\0' && *cursor == 'T' ) {
+    cursor++;
+    num = strtol(cursor, &rest, 10);
+    cursor = rest;
+  } else if (*cursor == '\0') {
+    return;
+  } else {
+    // error: malformed iso8601
+    return;
+  }
+
+  if (*cursor != '\0' && *cursor == 'H') {
+    _seconds += negative * SECS_PER_MIN * MINS_PER_HOUR * num;
+    cursor++;
+    num = strtol(cursor, &rest, 10);
+    cursor = rest;
+  }
+
+  if (*cursor != '\0' && *cursor == 'M') {
+    _seconds += negative * SECS_PER_MIN * num;
+    cursor++;
+    num = strtol(cursor, &rest, 10);
+    cursor = rest;
+  }
+
+  if (*cursor != '\0' && *cursor == 'S') {
+    _seconds += negative * num;
+  }
+}
+
+String TimeSpan::toString() const {
+  constexpr size_t buflen = 19;
+  char buf[buflen];
+  this->toCharArray(buf, buflen);
+  return String(buf);
+}
+
+/*!
+    @brief  Formats this timespan according to iso8601
+
+    Fails (returning 0) if buf is too small to store the result. buf size 19
+    will never fail.
+    Example: TimeSpan t(32, 23, 54, 11) formats to "P32DT23H54M11S".
+
+    @param buf char array to write output
+    @param len the length of buf
+    @return number of bytes written excluding trailing '\0' or 0 on failure
+*/
+size_t TimeSpan::toCharArray(char* buf, size_t len) const {
+  int32_t tmp = _seconds < 0 ? -_seconds: _seconds;
+  int8_t seconds = tmp % SECS_PER_MIN;
+  tmp /= SECS_PER_MIN;
+  int8_t minutes = tmp % MINS_PER_HOUR;
+  tmp /= MINS_PER_HOUR;
+  int8_t hours = tmp % HOURS_PER_DAY;
+  tmp /= HOURS_PER_DAY;
+  int16_t days = tmp;
+  size_t written = 0;
+  if (_seconds < 0) {
+    written += snprintf(buf+written, len-written, "-P");
```

Comment:
Note that some implementations of `snprintf()` return “the number of characters that would have been written if there were enough space”. This can make `written` larger than `len`.